### PR TITLE
Include ServerName in tls transport cache key

### LINKED ISSUE #56385: apiserver uses wrong CommonName to verify service certificates for aggregated API Server when External admission controller is enabled
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
After you configured webhook admission controller (ExternalAdmissionHookConfiguration) inside your cluster pointing to a service, apiserver will try to verify your new APIService certs against that admission controller service's Common Name. 
For example, a webhook admission controller is registered inside the cluster pointing to a service called "webhook" inside kube-system namespace (Common Name is webhook.kube-system.svc in this case). When you try to register a new APIService called `v1beta1.testapiservice.k8s.io` pointing to a service `testapiservice` inside `testns` namespace after that webhook admission controller was registered, you will see error inside apiserver log like this:
`E1127 01:22:06.230506       1 controller.go:111] loading OpenAPI spec for "v1beta1.testapiservice.k8s.io" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body: Error: 'x509: certificate is valid for testapiservice, testapiservice.testns, testapiservice.testns.svc, testapiservice.testns.svc.cluster.local, not webhook.kube-system.svc'`
And aggregation layer cannot talk to the custom apiserver due to invalid certs.

**What you expected to happen**:
apiserver should only verify the new aggregated API Server certs against its own service's Common Name and shouldn't verify it against webhook admission controller service's Common Name.

**How to reproduce it (as minimally and precisely as possible)**:
1. register a webhook admission controller inside your cluster like the example [here](https://github.com/caesarxuchao/example-webhook-admission-controller) using service webhook.kube-system.svc
2. register a new aggregated API Server using service testapiservice.testns.svc
3. aggregation layer cannot talk to the custom apiserver  and kube-apiserver generates error logs because it couldn't find valid Common Name `webhook.kube-system.svc` for your aggregated API Server service. 

**Anything else we need to know?**:
On the other hand admission controller works fine without any problem in this case described above.
We also tried deploying them in a different order. So deploy the aggregated API Server first and then register the admission controller webhook. In that case, the aggregated API server works fine but admission controller webhook won't work. From kube-apiserver log, you can see it's trying to verify the webhook admission controller service certs against aggregated API Server service's Common Name.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.3
- Cloud provider or hardware configuration: Aws
- OS (e.g. from /etc/os-release): Container Linux by CoreOS 1576.2.0 (Ladybug)
- Kernel (e.g. `uname -a`): 4.13.11-coreos
- Install tools:
- Others:

/sig api-machinery

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #56385 

```release-note
Fixes server name verification of aggregated API servers and webhook admission endpoints
```

### CODE REVIEW

**Comment by deads2k** [comment]
This will need picks.

/lgtm
/status approved-for-milestone

**Comment by liggitt** [comment]
picks opened

**Comment by liggitt** [comment]
integration test had a bug that was masked because the extension server and kube apiserver were both in-process and shared the tls config cache. pushed a test update

**Comment by deads2k** [comment]
still 
/lgtm

**Comment by caesarxuchao** [comment]
lgtm.

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 1]
```diff
@@ -88,5 +88,5 @@ func tlsConfigKey(c *Config) (string, error) {
 		return "", err
 	}
 	// Only include the things that actually affect the tls.Config
-	return fmt.Sprintf("%v/%x/%x/%x", c.TLS.Insecure, c.TLS.CAData, c.TLS.CertData, c.TLS.KeyData), nil
+	return fmt.Sprintf("%v/%x/%x/%x/%v", c.TLS.Insecure, c.TLS.CAData, c.TLS.CertData, c.TLS.KeyData, c.TLS.ServerName), nil
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/transport/cache.go]: Uh, are we printing very sensitive key data into the cache key? That seems like a terrible idea, doesn't it?

Comment (lavalamp) [reply staging/src/k8s.io/client-go/transport/cache.go]: It's safer to put things into a key struct, that way you don't have to worry about collisions (things maybe having `/` in them that shouldn't, say).

Comment (liggitt) [reply staging/src/k8s.io/client-go/transport/cache.go]: > Uh, are we printing very sensitive key data into the cache key? That seems like a terrible idea, doesn't it?

If you have memory access to the cache keys, you also have it to the cache data, which has the same information in a different form

> It's safer to put things into a key struct, that way you don't have to worry about collisions (things maybe having / in them that shouldn't, say).

`/` isn't an issue in this particular case with the current data (`%x` encodes, and the first value is a bool), but agree a key struct would be better here

Comment (lavalamp) [reply staging/src/k8s.io/client-go/transport/cache.go]: @liggitt 

> If you have memory access to the cache keys, you also have it to the cache data, which has the same information in a different form

I can easily see someone printing the keys to a debug log or something, which is readable by more people than they expect. People don't usually expect cache keys to be super sensitive.

I agree the separator is likely not an actual bug *in this particular case*, but if you use the struct approach everywhere you prevent an entire class of really hard to find bugs, some of which are serious security problems.

I will likely file an issue about these items.

Comment (liggitt) [reply staging/src/k8s.io/client-go/transport/cache.go]: fair enough. opened https://github.com/kubernetes/kubernetes/pull/56811 to switch to struct key with its own String() impl that omits logging the key data

Comment (lavalamp) [reply staging/src/k8s.io/client-go/transport/cache.go]: Thanks!

**Pull request merged**